design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,260 @@
+
+> **Note**: Exceptions to the period range may apply
+> if 2/3 of active maintainers come to agreement to prolong
+> a specific feature.
+
+### Deprecation
+One reason for features to go through the Alpha and Beta stages,
+is the opportunity to examine its usefulness and adoption by users.
+Same goes with major releases that intentionally break backward
+compatability.
+
+Therefore, it is only natural that some features will not graduate
+between the stages, or will be found irrelevant after some time and be
+discontinued when transitioning between major releases.
+
+This section will mainly focus on discontinuation of features between
+minor project releases.
+Discontinuation between major project releases is somehow simpler,
+as it is common and agreement (see semver) that backward compatability
+can be broken between major releases. It is worth noting that when a
+new major release is introduced, the previous one is still maintained
+and supported, something that does not exist with minor releases.
+
+#### The Deprecation Flow (for Minor releases)
+As stated before, only features that have not yet been graduated
+can be discontinued between minor releases.
+Therefore, a feature can be deprecated in Alpha and Beta stages only.
+
+These are the steps needed to deprecate a feature:
+- Proposal: Prepare a proposal to discontinue a feature with proper
+  reasoning, phases, exact timelines and functional alternatives (if any).
+  The proposal should be reviewed and approved.
+- Notification: Notify the project community about the feature
+  discontinuation based on the approved proposal.
+  All details of the plan should be provided to allow users and possibly
+  d/s projects to adjust.
+  Use all community media options to distribute this information
+  (e.g. mailing list, slack channel, community meetings).
+- Deprecation warnings: Add deprecation warnings at runtime to warn users
+  that the feature is planned to be discontinued.
+  Warnings should be raised when:
+  - Feature API fields are accessed.
+  - Feature FG is activated.
+  - Behavior related to the feature is detected (optional).
+- Removal: Feature removal involves dropping the core functionality
+  of a feature and its exposed API.
+  - The core implementation can be dropped in two steps:
+    - The FG is dropped by assuring it is never reported as set
+      (i.e. even if it is left by the operator configured, internally
+       it is ignored).
+      At this stage, the core implementation will follow the FG conditions
+      and therefore from the outside the feature is inactive.
+    - In case there are no side effects, the core implementation code can
+      be removed.
+  - The API types are not to be removed, as it may have implications
+    with the underlying storage which has already persisted them.
+    Kubernetes has not removed fields, it just kept them around with the
+    warning that they have been deprecated and no longer available.
```

Comment:
For Alpha features can we assume that no long living cluster will have used them? Because if we assume this, then we can remove these fields.

For Beta and GA features indeed, it has to be assumed that they were used, and we can not remove these fields anymore.

Reply:
Would this field still be exposed with OpenAPI?
Aka will it remain in the docs?

Reply:
We need to address this. Alpha/Beta fields must not be in the docs, should not be in the API.

Reply:
Ok, can we resovle the conversation? If yes, then please do so :)